kaldb/src/main/java/com/slack/kaldb/clusterManager/ClusterHpaMetricService.java

```diff
@@ -0,0 +1,172 @@
+package com.slack.kaldb.clusterManager;
+
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.slack.kaldb.metadata.cache.CacheSlotMetadataStore;
+import com.slack.kaldb.metadata.hpa.HpaMetricMetadata;
+import com.slack.kaldb.metadata.hpa.HpaMetricMetadataStore;
+import com.slack.kaldb.metadata.replica.ReplicaMetadata;
+import com.slack.kaldb.metadata.replica.ReplicaMetadataStore;
+import com.slack.kaldb.proto.metadata.Metadata;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The Cluster HPA metrics service is intended to be run on the manager node, and is used for making
+ * centralized, application-aware decisions that can be used to inform a Kubernetes horizontal pod
+ * autoscaler (HPA).
+ *
+ * @see <a
+ *     href="https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/">Kubernetes
+ *     HPA</a>
+ */
+public class ClusterHpaMetricService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(ClusterHpaMetricService.class);
+
+  // todo - consider making over-provision and lock duration configurable values
+  protected int CACHE_SLOT_OVER_PROVISION = 1000;
```

Comment:
so for each replica set we add 1000 to the totalReplicas to calculate demand. Is this value too high?

Reply:
This adds 1000 slots, which would be only a handful of instances depending on the slot density. We typically run between 200 - 500 slots, so that's 2-5 pods extra.

This will eventually probably be a config, but we would ideally target a number slightly above x2 the slot capacity so that any one instance can go down and we'd still have capacity for rollovers.